internal/client/client.go

```diff
@@ -298,15 +298,20 @@ func (c *Client) handleClusterEvent(event event.Event) {
 		return
 	}
 	c.Logger.Debug(func() string { return "cluster disconnected, rebooting" })
-	// try to reboot cluster connection
-	c.ConnectionManager.Stop()
-	c.ClusterService.Reset()
-	c.PartitionService.Reset()
-	if err := c.ConnectionManager.Start(ctx); err != nil {
-		c.Logger.Errorf("cannot reconnect to cluster, shutting down: %w", err)
-		// Shutdown is blocking operation which will make sure all the event goroutines are closed.
-		// If we wait here blocking, it will be a deadlock
-		go c.Shutdown(ctx)
+	for {
+		if c.State() != Ready {
+			break
+		}
+		// try to reboot cluster connection
+		c.ConnectionManager.Stop()
+		c.ClusterService.Reset()
+		c.PartitionService.Reset()
+		if err := c.ConnectionManager.Start(ctx); err != nil {
```

Comment:
Does Start() call being successful ensure we are connected to the cluster?

Reply:
That's right.